flexmeasures/data/migrations/versions/f0ee99278f6f_added_flexmodel_to_generic_asset_model.py

```diff
@@ -0,0 +1,308 @@
+
+            # check if value is a int, bool, float or dict
+            if not isinstance(flex_model_data[old_name], (int, float, dict, bool)):
+                raise Exception(
+                    f"Invalid value for '{old_name}' in generic asset {asset_id}: {flex_model_data[old_name]}"
+                )
+
+            if old_name[-6:] == "in_mwh":
+                # convert from float (in MWh) to string (in kWh)
+                value_in_kwh = flex_model_data[old_name] * 1000
+                flex_model_data[new_name] = f"{value_in_kwh} kWh"
+            elif old_name[-6:] == "in_mw":
+                # convert from float (in MW) to string (in kW)
+                value_in_kw = flex_model_data[old_name] * 1000
+                flex_model_data[new_name] = f"{value_in_kw} kW"
+            else:
+                # move as is
+                value = flex_model_data[old_name]
+                flex_model_data[new_name] = value
+
+            # Update the generic asset attributes to remove 'old_name' and add 'new_name' to flex_model
+            asset_attr_flex_model.pop(new_name, None)
+            asset_attr.pop(old_name)
+            flex_model_data.pop(old_name)
+            flex_model_data.pop("flex-model", None)
+            stmt = (
+                generic_asset_table.update()
+                .where(generic_asset_table.c.id == asset_id)
+                .values(
+                    flex_model=flex_model_data,
+                    attributes=asset_attr,
+                )
+            )
+            conn.execute(stmt)
+
+        # Process the grouped sensors
+        for sensor_generic_asset, sensors_with_key in field_spec["grouped"].items():
+            sensor = sensors_with_key[0]
+            field_value = None
+
+            if sensor.attributes.get(old_name) is not None:
+                field_value = sensor.attributes.get(old_name)
+            elif sensor_generic_asset.attributes.get(old_name) is not None:
+                field_value = sensor_generic_asset.attributes.get(old_name)
+
+            if field_value is not None:
+                # check if value is a int or float
+                if not isinstance(field_value, (int, float)):
+                    raise Exception(
+                        f"Invalid value for '{old_name}' in sensor {sensor.id}: {sensor.attributes[old_name]}"
+                    )
+
+                soc_min_value_kwh = field_value * 1000
+                soc_min_in_kwh = f"{soc_min_value_kwh} kWh"
+                flex_model_data = {new_name: soc_min_in_kwh}
+
+                # Update the generic asset attributes to remove 'old_name' and add 'new_name' to flex_model
+                sensor_generic_asset.attributes.pop(old_name, None)
+                stmt = (
+                    generic_asset_table.update()
+                    .where(generic_asset_table.c.id == asset_id)
+                    .values(
+                        flex_model=flex_model_data,
```

Comment:
Is this not potentially overwriting existing `flex_model` data on that asset, which the previous loop, or a previous run in this loop might have set?

Line 192 might have the same problem.

Or are you sure you are updating each asset exactly once? I can't tell.